### PR TITLE
Feature merge

### PR DESCRIPTION
**Pull request by SalusaSecondus**
This pull-request is a roll-up to synchronize functionality and release several minor features.

**Together these changes are sufficiently important that a minor version bump to 1.12.0 is included in this PR.**

Changes:
* Minor version change to 1.12.0
* Dependency versions are updated
* Added `MetaStore.replicate` to make it easier to copy data-keys from one MetaStore to another
* Better support for non-standard JCA providers
* Fixed thread-safety issue in LRUCache

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java [hunk 1]
```diff
@@ -371,6 +380,10 @@ private void actualDecryption(Map<String, AttributeValue> itemAttributes,
         }
     }
 
+    protected int getBlockSize(final String encryptionMode) {
```
Comment (bdonlan) [inline src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java]: Should be static.

File: src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java [hunk 2]
```diff
@@ -405,16 +418,22 @@ private void actualEncryption(Map<String, AttributeValue> itemAttributes,
...
                     }
                     // Encryption format: <iv><ciphertext>
                     // Note a unique iv is generated per attribute
-                    byte[] iv = Utils.getRandom(ivSize);
-                    cipher.init(Cipher.ENCRYPT_MODE, encryptionKey, new IvParameterSpec(iv), Utils.getRng());
-                    cipherText = ByteBuffer.allocate(ivSize + cipher.getOutputSize(plainText.remaining()));
-                    cipherText.put(iv);
+                    cipher.init(Cipher.ENCRYPT_MODE, encryptionKey, Utils.getRng());
+                    cipherText = ByteBuffer.allocate(blockSize + cipher.getOutputSize(plainText.remaining()));
+                    cipherText.position(blockSize);
```
Comment (bdonlan) [inline src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java]: Block size may not equal IV size, depending on the mode (e.g. AES-GCM would use a 12 byte IV with a 16 byte block size.

File: src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java [hunk 3]
```diff
@@ -539,17 +558,22 @@ public String getSigningAlgorithmHeader() {
...
+        if (buffer.hasArray() && buffer.arrayOffset() == 0) {
             byte[] result = buffer.array();
-            buffer.rewind();
-            return result;
-        } else {
-            byte[] result = new byte[buffer.remaining()];
-            buffer.get(result);
-            buffer.rewind();
-            return result;
+            if (buffer.remaining() == result.length) {
```
Comment (bdonlan) [inline src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java]: Combine into the if condition.

File: src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/internal/LRUCache.java [hunk 4]
```diff
@@ -98,11 +99,16 @@ public void clear() {
...
+            List<Entry<String, T>> removedEntries = new ArrayList<Entry<String, T>>();
+            synchronized (map) {
+                Iterator<Entry<String, T>> it = map.entrySet().iterator();
+                while(it.hasNext()) {
+                    removedEntries.add(it.next());
+                    it.remove();
+                }
+            }
+            for (Entry<String, T> entry : removedEntries) {
+                listener.onRemoval(entry);
```
Comment (bdonlan) [inline src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/internal/LRUCache.java]: Does anything depend on the synchronization we're removing here? It's now possible to have onRemove called for the same key K twice, with a newer version being called before the older one.

File: src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java (hunk 2 above)
Comment (SalusaSecondus) [reply src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java]: In this data-format  version of the client they must be the same. It's a limitation of the format which we cannot change until we do a more significant version rev.

File: src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java (hunk 3 above)
Comment (SalusaSecondus) [reply src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java]: Can't do as `result` can only be safely retrieved from `buffer.array()` after the first conditional and we want to check it for sanity prior to returning it.

File: src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java (hunk 2 above)
Comment (bdonlan) [reply src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java]: Please leave a comment to this effect, and a runtime check to confirm that the returned `iv` array is the correct size.

File: src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/internal/LRUCache.java (hunk 4 above)
Comment (bdonlan) [reply src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/internal/LRUCache.java]: Ah, never mind, I thought it was moving the onRemoval outside of the synchronized block, but now I see there wasn't synchronization before.

File: src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java (hunk 3 above)
Comment (bdonlan) [reply src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/encryption/DynamoDBEncryptor.java]: got it, the `byte[] result` assignment was lost between two diff blocks.

**Pull request merged**